### PR TITLE
[Converter] Combined Converters

### LINKED ISSUE #553: [Enhancement] Converters modification
## Summary

There are converters which have return the value and the opposite value. They can be combined into 1 converter by using converter parameter.
e.g. IsNullOrEmptyConverter vs IsNotNullOrEmptyConverter. ListIsEmptyConverter vs ListIsNotEmptyConverter

## API Changes

1. remove *Not*Converters
2. use boolean parameters.

## Who Will Do The Work?

- [ ] I am willing to take this on myself
- [ ] Just putting this out there for someone to pick up

### PR DESCRIPTION
**Pull request by GUOLDEV**
Fixes #553 

Combined Converters IsNotNullOrEmptyConverter, ListIsNullOrEmptyConverter and ListIsNotNullOrEmptyConverter into IsNullOrEmptyConverter, implementing InvertCheck flag to invert the result.

If it is ok, IsNotNullOrEmptyConverter, ListIsNullOrEmptyConverter and ListIsNotNullOrEmptyConverter can be removed as all of them can be replaced using IsNullOrEmptyConverter with InvertCheck flag.

### Bugs Fixed ###

NA

### API Changes ###

<!-- List all API changes here (or just put None), example:

Added: 
 
- `string Class.Property { get; set; }`
- `void Class.Method();`

Changed:

 - `object Cell.OldPropertyName` => `object Cell.NewPropertyName`
 
If there is an entirely new API, then you can use a more verbose style:

```csharp
public static class NewClass {
    public static int SomeProperty { get; set; }
    public static void SomeMethod(string value);
}
``` -->
Changed:

- IsNullOrEmptyConverter
- IsNotNullOrEmptyConverter
- ListIsNullOrEmptyConverter
- ListIsNotNullOrEmptyConverter

### Behavioral Changes ###

<!-- Describe any non-bug related behavioral changes that may change how users app behaves when upgrading to this version of the codebase. -->

NA

### PR Checklist ###
<!-- Please check all the things you did here and double-check that you got it all, or state why you didn't do something -->

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [ ] Changes adhere to coding standard
<!-- If at all possible, please update/add the documentation on the repo below. We would very much appreciate that. If you are unable to, please consider at least opening an issue on the repo below so we know that Docs still need to be adjusted/created. Thanks! <3 -->
- [ ] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

**Comment by GUOLDEV** [comment]
@Cfun1 thanks for the feedback, the PR targeting develop branch.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs [hunk 1]
```diff
@@ -18,10 +24,25 @@ public class IsNullOrEmptyConverter : ValueConverterExtension, IValueConverter
...
-		public object Convert(object? value, Type? targetType, object? parameter, CultureInfo? culture) => ConvertInternal(value);
+		public object Convert(object? value, Type? targetType, object? parameter, CultureInfo? culture)
+			=> InvertCheck != ConvertInternal(value);
+
+		internal static bool ConvertInternal(object? value, bool isListCheck = false)
+		{
+			if (value == null)
+				return true;
+
+			if (value is IEnumerable list)
```
Comment (maxkoshevoi) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs:35]: String is an array of `chars`, so this `if` will be true for all strings. 
1) This means that  `value is string str` bellow is redundant
2) This method will return `false` for string that contains whitespace, while previously it returned `true` in such cases

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs [hunk 2]
```diff
@@ -10,6 +11,11 @@ namespace Xamarin.CommunityToolkit.Converters
 	/// </summary>
 	public class IsNullOrEmptyConverter : ValueConverterExtension, IValueConverter
 	{
+		/// <summary>
+		/// A value that specifies whether or not invert the result.
+		/// </summary>
+		public bool InvertCheck { get; set; }
```
Comment (maxkoshevoi) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs:17]: I think better name would be simply `Invert` or `InvertResult`

File: src/CommunityToolkit/Xamarin.CommunityToolkit.UnitTests/Converters/IsNullOrEmptyConverter_Tests.cs [hunk 3]
```diff
@@ -1,15 +1,25 @@
...
+		public static IEnumerable<object?[]> GetData() => new List<object?[]>
+		{
+			new object?[] { null, true },
+			new object[] { string.Empty, true },
+			new object[] { "Test", false },
+			new object[] { typeof(IsNullOrEmptyConverter), false },
+			new object[] { new List<string>(), true },
+			new object[] { new List<string>() { "TestValue" }, false },
+			new object[] { Enumerable.Range(1, 3), false },
+		};
```
Comment (maxkoshevoi) [inline src/CommunityToolkit/Xamarin.CommunityToolkit.UnitTests/Converters/IsNullOrEmptyConverter_Tests.cs:20]: Add test for whitespace string:
`new object[] { " ", true },`

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs [hunk 4]
```diff
@@ -18,10 +24,25 @@ public class IsNullOrEmptyConverter : ValueConverterExtension, IValueConverter
 		/// <param name="parameter">Additional parameter for the converter to handle. This is not implemented.</param>
 		/// <param name="culture">The culture to use in the converter. This is not implemented.</param>
 		/// <returns>A <see cref="bool"/> indicating if the incoming value is null or empty.</returns>
-		public object Convert(object? value, Type? targetType, object? parameter, CultureInfo? culture) => ConvertInternal(value);
+		public object Convert(object? value, Type? targetType, object? parameter, CultureInfo? culture)
+			=> InvertCheck != ConvertInternal(value);
```
Comment (maxkoshevoi) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs:28]: It's not obvious (at least to me) that `InvertCheck !=` inverts the result of `ConvertInternal`. I think inversion should be written more clearly for better readability.

```suggestion
		{
			var result = ConvertInternal(value);
			if (InvertCheck)
				result = !result;
			
			return result;
		}
```

**Review by maxkoshevoi — COMMENTED** [review]
Thank you for the PR. Great idea!

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs (hunk 1 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs:44]: Done!

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs (hunk 2 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs]: Done!

File: src/CommunityToolkit/Xamarin.CommunityToolkit.UnitTests/Converters/IsNullOrEmptyConverter_Tests.cs (hunk 3 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit.UnitTests/Converters/IsNullOrEmptyConverter_Tests.cs:21]: Done!

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs (hunk 4 above)
Comment (GUOLDEV) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/IsNullOrEmptyConverter.shared.cs]: Done!

**Comment by GUOLDEV** [comment]
Thanks for all the feedback!

**Comment by GUOLDEV** [comment]
Hey @maxkoshevoi , is there something else that I need to do here?

**Comment by pictos** [comment]
Closing this PR, as mentioned in this [issue](https://github.com/xamarin/XamarinCommunityToolkit/issues/553#issuecomment-816856684)
